stable_baselines3/common/pytree_dataclass.py

```diff
@@ -1,21 +1,282 @@
-from typing import Callable, TypeVar
+import dataclasses
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    overload,
+)
 
 import optree as ot
-from optree import PyTree as PyTree
+import torch as th
+from optree import CustomTreeNode, PyTree
+from typing_extensions import dataclass_transform
 
-__all__ = ["tree_flatten", "PyTree"]
+from stable_baselines3.common.type_aliases import TensorIndex
+from stable_baselines3.common.utils import zip_strict
+
+__all__ = [
+    "FrozenPyTreeDataclass",
+    "MutablePyTreeDataclass",
+    "TensorTree",
+    "tree_empty",
+    "tree_flatten",
+    "tree_index",
+    "tree_map",
+]
 
 T = TypeVar("T")
+U = TypeVar("U")
 
 SB3_NAMESPACE = "stable-baselines3"
 
+_RESERVED_NAMES = ["_PyTreeDataclassBase", "FrozenPyTreeDataclass", "MutablePyTreeDataclass"]
+
+
+# We need to inherit from `type(CustomTreeNode)` to prevent conflicts due to different-inheritance in metaclasses.
+# - For some reason just inheriting from `typing._ProtocolMeta` does not get rid of that error.
+# - Inheriting from `typing._GenericAlias` is impossible, as it's a `typing._Final` class.
+#
+# But in mypy, inheriting from a dynamic base class from `type` is not supported, so we disable type checking for this
+# line.
+class _PyTreeDataclassMeta(type(CustomTreeNode)):  # type: ignore[misc]
+    """Metaclass to register dataclasses as PyTrees.
+
+    Usage:
+      class MyDataclass(metaclass=_DataclassPyTreeMeta):
+        ...
+    """
+
+    # We need to have this `currently_registering` variable because, in the course of making a DataClass with __slots__,
+    # another class is created. So this will be called *twice* for every dataclass we annotate with this metaclass.
+    currently_registering: ClassVar[Optional[type]] = None
+
+    def __new__(mcs, name, bases, namespace, slots=True, **kwargs):
+        # First: create the class in the normal way.
+        cls = super().__new__(mcs, name, bases, namespace)
+
+        if dataclasses.is_dataclass(cls):
+            # If the class we're registering is already a Dataclass, it means it is a descendant of FrozenPyTreeDataclass or
+            # MutablePyTreeDataclass.
+            # This includes the children which are created when we create a dataclass with __slots__.
+
+            if mcs.currently_registering is not None:
+                # We've already created and annotated a class without __slots__, now we create the one with __slots__
+                # that will actually get returned after from the __new__ method.
+                assert mcs.currently_registering.__module__ == cls.__module__
+                assert mcs.currently_registering.__name__ == cls.__name__
+                mcs.currently_registering = None
+                return cls
+
+            else:
```

Comment:
This 'else' is unnecessary as the `if` always kicks out of the function (either through an assertion failure, or a direct return). Removing the 'else' makes the nesting a little less severe, which is IMO slightly nice for reading (and makes it so I don't have to think about which `if` it corresponds to).